In [1]:
import numpy as np
import pandas as pd

In [2]:
dict1 = {
    "Name":['kris', 'Ash', 'May', 'Dawn'], 
    "Age":[23,16,15,15],
    "City":['Kalos', 'Kento', 'Henoi', 'Sinnoh']
}

In [3]:
df = pd.DataFrame(dict1) #to construct a dataframe object
print(df)
#indexes are nothing but rows and headings are columns. indexes start from 0 by default and are appended infront of the dataframe

   Name  Age    City
0  kris   23   Kalos
1   Ash   16   Kento
2   May   15   Henoi
3  Dawn   15  Sinnoh


In [4]:
df.to_csv('friends.csv') #to convert a data frame into comma seperated values file and save it
df.to_csv('friends_index_false.csv', index = False) # to get rid of indices

In [5]:
#to view restricted rows in a data frame:
print(df.head(2),'\n') # first 2 rows/records
print(df.tail(3),'\n') # last 3 rows/records

   Name  Age   City
0  kris   23  Kalos
1   Ash   16  Kento 

   Name  Age    City
1   Ash   16   Kento
2   May   15   Henoi
3  Dawn   15  Sinnoh 



In [6]:
#to get stats of numerical columns:
print(df.describe(),'\n')

            Age
count   4.00000
mean   17.25000
std     3.86221
min    15.00000
25%    15.00000
50%    15.50000
75%    17.75000
max    23.00000 



In [7]:
#to read a csv file which is already present in the folder:
trains = pd.read_csv('trains.csv') 
print(trains)

   train number  speed       city
0         12221     35  ahmedabad
1         20112    120     mumbai
2         20010    360       pune
3         43332     86      delhi


In [8]:
#to access a column of a dataframe (like dictionary only)
print(trains['speed']) 


0     35
1    120
2    360
3     86
Name: speed, dtype: int64


In [9]:
#to access and change a value of the dataframe (like dictionary only)
# although this works, we should not change values in dataframe like this. We will discuss the reason later on.
trains['speed'][0] = 200
print(trains,'\n')

   train number  speed       city
0         12221    200  ahmedabad
1         20112    120     mumbai
2         20010    360       pune
3         43332     86      delhi 



C:\Users\HARSH\AppData\Local\Temp\ipykernel_7972\959153449.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  trains['speed'][0] = 200


In [10]:
#to update the csv file in the folder:
# trains.to_csv('trains.csv')

In [11]:
#to change the indexing format:
trains.index = ['first', 'second', 'third', 'fourth']
print(trains)

        train number  speed       city
first          12221    200  ahmedabad
second         20112    120     mumbai
third          20010    360       pune
fourth         43332     86      delhi


In [12]:
print(trains['city']['second']) #now you can use this indexing to get rows

mumbai


###### PANDAS SERIES IS A ONE DIMENTIONAL DATA STRUCTURE (sort of like column vector) WHICH CAN HOLD DATA OF ANY TYPE


In [13]:
ser = pd.Series(np.random.rand(20))
print(ser,'\n')
print(type(ser),'\n')
print(ser[5],'\n')

0     0.691205
1     0.426908
2     0.488493
3     0.671710
4     0.256683
5     0.135542
6     0.212652
7     0.771439
8     0.130556
9     0.546165
10    0.155429
11    0.269074
12    0.692624
13    0.594585
14    0.112199
15    0.216165
16    0.424884
17    0.490368
18    0.536928
19    0.817414
dtype: float64 

<class 'pandas.core.series.Series'> 

0.13554209227425407 



###### PANDAS DATAFRAME IS A TWO DIMENTIONAL DATA STRUCTURE WHICH CAN HOLD MULTIPLE SERIES OR ARRAYS OR COLUMNS WITH DIFFERENT DATA TYPE ELEMENTS.


In [14]:
newdf = pd.DataFrame(np.random.rand(20,3), index = np.arange(20), columns=np.arange(1,4,1))
print(newdf)
print(type(newdf),'\n')
print(newdf.dtypes,'\n')  #its 'dtypes' not 'dtype'!


           1         2         3
0   0.997756  0.877487  0.401267
1   0.221990  0.271778  0.921321
2   0.870535  0.793812  0.093922
3   0.651911  0.235065  0.292353
4   0.757720  0.057464  0.402509
5   0.312491  0.242250  0.850516
6   0.904044  0.517113  0.590413
7   0.948534  0.177507  0.876713
8   0.711115  0.926866  0.239869
9   0.450242  0.218081  0.098219
10  0.426621  0.280005  0.431881
11  0.907079  0.537089  0.307211
12  0.621882  0.459242  0.679232
13  0.624936  0.714154  0.789407
14  0.567130  0.230058  0.562841
15  0.584737  0.194991  0.548481
16  0.172280  0.302154  0.715641
17  0.916317  0.822276  0.781340
18  0.489627  0.481115  0.741343
19  0.242591  0.600438  0.332537
<class 'pandas.core.frame.DataFrame'> 

1    float64
2    float64
3    float64
dtype: object 



In [15]:
newdf[1][0] = 'clay'
print(newdf.head(),'\n') #if nothing in head parameter then by default it shows 5 rows
print(newdf.dtypes) #note now the 0th index column has data type of object since it has both float and string as elements


          1         2         3
0      clay  0.877487  0.401267
1   0.22199  0.271778  0.921321
2  0.870535  0.793812  0.093922
3  0.651911  0.235065  0.292353
4   0.75772  0.057464  0.402509 

1     object
2    float64
3    float64
dtype: object


In [16]:
print(newdf.index,'\n')
print(newdf.columns,'\n')

Int64Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,
            19],
           dtype='int64') 

Int64Index([1, 2, 3], dtype='int64') 



In [17]:
df.columns

Index(['Name', 'Age', 'City'], dtype='object')

In [18]:
newdf[1][0] = 0.8976372242424
print(newdf.to_numpy(),'\n') # to convert dataframe into numpy array

[[0.8976372242424 0.8774865928599472 0.4012671941939002]
 [0.22198972174677878 0.27177804367283964 0.9213209908288454]
 [0.8705348562565212 0.7938121876837585 0.09392163061442427]
 [0.6519107471816206 0.23506452455918359 0.29235325954383373]
 [0.7577201512320351 0.05746352285801615 0.402509145822771]
 [0.3124907532979201 0.2422500954606981 0.8505161874612011]
 [0.9040435129513175 0.5171128601753111 0.5904129720985458]
 [0.9485338247517606 0.17750732264496594 0.8767133985538875]
 [0.7111150927946513 0.9268661949019225 0.23986903706129603]
 [0.45024230448547586 0.21808129958601252 0.09821927479046633]
 [0.4266212749328886 0.2800050441029541 0.43188145385162224]
 [0.9070789413988041 0.537089467755647 0.3072109870604681]
 [0.6218820498942935 0.4592419386315869 0.6792322082339134]
 [0.6249355644259224 0.7141536600437628 0.7894067190363344]
 [0.5671301487370104 0.2300581024950955 0.5628414341365963]
 [0.5847372496142514 0.19499097846739888 0.5484812062733698]
 [0.17227952218954234 0.30215421

In [19]:
print(newdf.T,'\n') #to transpose the dataframe

         0         1         2         3         4         5         6   \
1  0.897637   0.22199  0.870535  0.651911   0.75772  0.312491  0.904044   
2  0.877487  0.271778  0.793812  0.235065  0.057464   0.24225  0.517113   
3  0.401267  0.921321  0.093922  0.292353  0.402509  0.850516  0.590413   

         7         8         9         10        11        12        13  \
1  0.948534  0.711115  0.450242  0.426621  0.907079  0.621882  0.624936   
2  0.177507  0.926866  0.218081  0.280005  0.537089  0.459242  0.714154   
3  0.876713  0.239869  0.098219  0.431881  0.307211  0.679232  0.789407   

         14        15        16        17        18        19  
1   0.56713  0.584737   0.17228  0.916317  0.489627  0.242591  
2  0.230058  0.194991  0.302154  0.822276  0.481115  0.600438  
3  0.562841  0.548481  0.715641   0.78134  0.741343  0.332537   



In [20]:
print(newdf.sort_index(axis = 0, ascending = False),'\n')  #axis = 0 so it's for rows 
print(newdf.sort_index(axis = 1, ascending = False),'\n')  #axis = 1 so it's for cols


           1         2         3
19  0.242591  0.600438  0.332537
18  0.489627  0.481115  0.741343
17  0.916317  0.822276  0.781340
16   0.17228  0.302154  0.715641
15  0.584737  0.194991  0.548481
14   0.56713  0.230058  0.562841
13  0.624936  0.714154  0.789407
12  0.621882  0.459242  0.679232
11  0.907079  0.537089  0.307211
10  0.426621  0.280005  0.431881
9   0.450242  0.218081  0.098219
8   0.711115  0.926866  0.239869
7   0.948534  0.177507  0.876713
6   0.904044  0.517113  0.590413
5   0.312491  0.242250  0.850516
4    0.75772  0.057464  0.402509
3   0.651911  0.235065  0.292353
2   0.870535  0.793812  0.093922
1    0.22199  0.271778  0.921321
0   0.897637  0.877487  0.401267 

           3         2         1
0   0.401267  0.877487  0.897637
1   0.921321  0.271778   0.22199
2   0.093922  0.793812  0.870535
3   0.292353  0.235065  0.651911
4   0.402509  0.057464   0.75772
5   0.850516  0.242250  0.312491
6   0.590413  0.517113  0.904044
7   0.876713  0.177507  0.948534
8   0.23

In [21]:
print(newdf[1],'\n')  # NOTE THAT THIS WILL GIVE COLUMN NOT ROW!! WHICH HAPPENS GENEREALLY IN ANY LIST OR NP ARRAY.
print(type(newdf[1]),'\n')
# SO COMBINATION OF SERIES MAKES A DATAFRAME

0     0.897637
1      0.22199
2     0.870535
3     0.651911
4      0.75772
5     0.312491
6     0.904044
7     0.948534
8     0.711115
9     0.450242
10    0.426621
11    0.907079
12    0.621882
13    0.624936
14     0.56713
15    0.584737
16     0.17228
17    0.916317
18    0.489627
19    0.242591
Name: 1, dtype: object 

<class 'pandas.core.series.Series'> 



###### DIFFERENCE BETWEEN COPY AND VIEW: 

In [22]:
newdf2 = newdf
#newdf2 is a view of newdf, not a copy of it! hence both are pointing at the same values just like we had in python and numpy
#so if i change anything in newdf2, it will change newdf also
newdf2[1][1] = 100
print(newdf2,'\n')
print(newdf,'\n')
# see both got changed

           1         2         3
0   0.897637  0.877487  0.401267
1        100  0.271778  0.921321
2   0.870535  0.793812  0.093922
3   0.651911  0.235065  0.292353
4    0.75772  0.057464  0.402509
5   0.312491  0.242250  0.850516
6   0.904044  0.517113  0.590413
7   0.948534  0.177507  0.876713
8   0.711115  0.926866  0.239869
9   0.450242  0.218081  0.098219
10  0.426621  0.280005  0.431881
11  0.907079  0.537089  0.307211
12  0.621882  0.459242  0.679232
13  0.624936  0.714154  0.789407
14   0.56713  0.230058  0.562841
15  0.584737  0.194991  0.548481
16   0.17228  0.302154  0.715641
17  0.916317  0.822276  0.781340
18  0.489627  0.481115  0.741343
19  0.242591  0.600438  0.332537 

           1         2         3
0   0.897637  0.877487  0.401267
1        100  0.271778  0.921321
2   0.870535  0.793812  0.093922
3   0.651911  0.235065  0.292353
4    0.75772  0.057464  0.402509
5   0.312491  0.242250  0.850516
6   0.904044  0.517113  0.590413
7   0.948534  0.177507  0.876713
8   0.71

In [23]:
#you can prevent this using copy() method 
newdf3 = newdf.copy()  # OR USE NEWDF3 = NEWDF[:] FOR COPYING
newdf3[1][1] = 200
print(newdf3,'\n')
print(newdf,'\n')

           1         2         3
0   0.897637  0.877487  0.401267
1        200  0.271778  0.921321
2   0.870535  0.793812  0.093922
3   0.651911  0.235065  0.292353
4    0.75772  0.057464  0.402509
5   0.312491  0.242250  0.850516
6   0.904044  0.517113  0.590413
7   0.948534  0.177507  0.876713
8   0.711115  0.926866  0.239869
9   0.450242  0.218081  0.098219
10  0.426621  0.280005  0.431881
11  0.907079  0.537089  0.307211
12  0.621882  0.459242  0.679232
13  0.624936  0.714154  0.789407
14   0.56713  0.230058  0.562841
15  0.584737  0.194991  0.548481
16   0.17228  0.302154  0.715641
17  0.916317  0.822276  0.781340
18  0.489627  0.481115  0.741343
19  0.242591  0.600438  0.332537 

           1         2         3
0   0.897637  0.877487  0.401267
1        100  0.271778  0.921321
2   0.870535  0.793812  0.093922
3   0.651911  0.235065  0.292353
4    0.75772  0.057464  0.402509
5   0.312491  0.242250  0.850516
6   0.904044  0.517113  0.590413
7   0.948534  0.177507  0.876713
8   0.71

C:\Users\HARSH\AppData\Local\Temp\ipykernel_7972\1751267931.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  newdf3[1][1] = 200


###### THIS IS ALSO THE REASON WHY WE GOT THE WARNING OF SettingWithCopy when we tried to change the value of a dataframe element. when we are chaining (like doing newdf[1][0] = 100) then it is not a guarranty that we will get a view or a copy so we don't set a value like this, newdf[1][0] = 100 rather, we use loc and iloc methods. Always set values using loc or iloc!!!! loc and iloc are used to set as well as retrieve data.


In [24]:
newdf.loc[0,1] = 300 # first parameter for row, second for column like usual np array 
print(newdf.head(2),'\n') 

     1         2         3
0  300  0.877487  0.401267
1  100  0.271778  0.921321 



In [25]:
newdf.columns = list('ABC') # converted the string 'ABC' into a list and then that list is allocated as column names
print(newdf.head(2),'\n')

     A         B         C
0  300  0.877487  0.401267
1  100  0.271778  0.921321 



In [26]:
newdf.loc[0,'A'] = 500
print(newdf.head(2),'\n')

     A         B         C
0  500  0.877487  0.401267
1  100  0.271778  0.921321 



In [27]:
newdf.loc[0,0] = 99 # this will create a new column with name 0 ! and rest of the values will be NAN
print(newdf.head(2),'\n')

     A         B         C     0
0  500  0.877487  0.401267  99.0
1  100  0.271778  0.921321   NaN 



In [28]:
newdf = newdf.drop([0], axis = 1) # it will drop the 0 named column (cause axis = 1)
print(newdf.head(2),'\n')

     A         B         C
0  500  0.877487  0.401267
1  100  0.271778  0.921321 



In [29]:
# getting elements from dataframe:
print(newdf.loc[[1,2,4],['B','C']],'\n') # this is only a view, hence inside the print. newdf is still intact. if you wanna change newdf then make newdf equal to this

          B         C
1  0.271778  0.921321
2  0.793812  0.093922
4  0.057464  0.402509 



In [30]:
print(newdf.loc[:, ['B','C']],'\n') # to get all rows

           B         C
0   0.877487  0.401267
1   0.271778  0.921321
2   0.793812  0.093922
3   0.235065  0.292353
4   0.057464  0.402509
5   0.242250  0.850516
6   0.517113  0.590413
7   0.177507  0.876713
8   0.926866  0.239869
9   0.218081  0.098219
10  0.280005  0.431881
11  0.537089  0.307211
12  0.459242  0.679232
13  0.714154  0.789407
14  0.230058  0.562841
15  0.194991  0.548481
16  0.302154  0.715641
17  0.822276  0.781340
18  0.481115  0.741343
19  0.600438  0.332537 



In [31]:
print(newdf.loc[[1,2,4], :],'\n') # to get all columns

          A         B         C
1       100  0.271778  0.921321
2  0.870535  0.793812  0.093922
4   0.75772  0.057464  0.402509 



###### some advanced queries with masking:

In [32]:
# making the mask:
((newdf["A"] < 0.3) | (newdf['C'] > 0.7)) #remember first argument is column in dataframe not row.

0     False
1      True
2     False
3     False
4     False
5      True
6     False
7      True
8     False
9     False
10    False
11    False
12    False
13     True
14    False
15    False
16     True
17     True
18     True
19     True
dtype: bool

In [33]:
print(newdf.loc[((newdf["A"] < 0.3) | (newdf['C'] > 0.7))],'\n') #remember first argument is column in dataframe not row.
print(newdf.loc[((newdf["A"] < 0.3) & (newdf['C'] > 0.7))],'\n')


           A         B         C
1        100  0.271778  0.921321
5   0.312491  0.242250  0.850516
7   0.948534  0.177507  0.876713
13  0.624936  0.714154  0.789407
16   0.17228  0.302154  0.715641
17  0.916317  0.822276  0.781340
18  0.489627  0.481115  0.741343
19  0.242591  0.600438  0.332537 

          A         B         C
16  0.17228  0.302154  0.715641 



###### DIFFERENCE BETWEEN LOC AND ILOC: IN LOC YOU HAVE TO USE THE NAMES GIVEN TO ROWS AND COLS IN ORDER TO ACCESS THEM. IN ILOC YOU HAVE TO USE THE 0 INDEXING TO ACCESS BOTH ROWS AND COLS IRRESPECTIVE OF THE NAMES OF THE ROWS AND COLS


In [34]:
print(newdf.iloc[:, 1],'\n') #this will give nothing by 'B' column
print(newdf.loc[:, 'B'],'\n') # see same output as above
 

0     0.877487
1     0.271778
2     0.793812
3     0.235065
4     0.057464
5     0.242250
6     0.517113
7     0.177507
8     0.926866
9     0.218081
10    0.280005
11    0.537089
12    0.459242
13    0.714154
14    0.230058
15    0.194991
16    0.302154
17    0.822276
18    0.481115
19    0.600438
Name: B, dtype: float64 

0     0.877487
1     0.271778
2     0.793812
3     0.235065
4     0.057464
5     0.242250
6     0.517113
7     0.177507
8     0.926866
9     0.218081
10    0.280005
11    0.537089
12    0.459242
13    0.714154
14    0.230058
15    0.194991
16    0.302154
17    0.822276
18    0.481115
19    0.600438
Name: B, dtype: float64 



In [35]:
print(newdf.drop(['A','C'], axis = 1),'\n') #this won't change the dataframe, it's only for view. if you wanna change the make dfnew equal to this.
newdf.drop(['A','C'], axis = 1, inplace = True) #this will overide the dataframe. since inplace is true.
print(newdf,'\n')


           B
0   0.877487
1   0.271778
2   0.793812
3   0.235065
4   0.057464
5   0.242250
6   0.517113
7   0.177507
8   0.926866
9   0.218081
10  0.280005
11  0.537089
12  0.459242
13  0.714154
14  0.230058
15  0.194991
16  0.302154
17  0.822276
18  0.481115
19  0.600438 

           B
0   0.877487
1   0.271778
2   0.793812
3   0.235065
4   0.057464
5   0.242250
6   0.517113
7   0.177507
8   0.926866
9   0.218081
10  0.280005
11  0.537089
12  0.459242
13  0.714154
14  0.230058
15  0.194991
16  0.302154
17  0.822276
18  0.481115
19  0.600438 



In [36]:
newdf.drop([1,4], axis = 0, inplace = True) # dropping rows
print(newdf.head(),'\n')

          B
0  0.877487
2  0.793812
3  0.235065
5  0.242250
6  0.517113 



In [37]:
# to reset the index after you have dropped some rows:
newdf.reset_index(drop = True, inplace = True) # drop=true to get rid of that index column of previous indexes
print(newdf.head(),'\n')


          B
0  0.877487
1  0.793812
2  0.235065
3  0.242250
4  0.517113 



In [38]:
newdf = newdf.reset_index() #to have one more col as previous indexes
print(newdf.head(),'\n')

   index         B
0      0  0.877487
1      1  0.793812
2      2  0.235065
3      3  0.242250
4      4  0.517113 



In [39]:
newdf.columns = ['previous index', 'B']
print(newdf.head(),'\n')

   previous index         B
0               0  0.877487
1               1  0.793812
2               2  0.235065
3               3  0.242250
4               4  0.517113 



In [40]:
newdf.loc[:, 'previous index'] = 0
print(newdf.head(),'\n')

   previous index         B
0               0  0.877487
1               0  0.793812
2               0  0.235065
3               0  0.242250
4               0  0.517113 



In [41]:
print(newdf['previous index'].isnull(),'\n') # all false because they have some value (0 in this case)

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
Name: previous index, dtype: bool 



In [42]:
newdf.loc[0:5, 'previous index'] = None
print(newdf['previous index'].isnull(),'\n') # true when some values are not present or null or None or NaN or blank


0      True
1      True
2      True
3      True
4      True
5      True
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
Name: previous index, dtype: bool 



In [43]:
newestdf = pd.DataFrame({
    "name": ['Alfred', 'Batman', 'Catwoman'],
    "toy": [np.nan, 'Batmobile', 'Bullwhip'],
    "born": [pd.NaT, pd.Timestamp("1940-04-25"), pd.NaT]
    }
)
print(newestdf,'\n') # pd.NaT is missing date. np.nan is not a number or null/None value or missing value.


       name        toy       born
0    Alfred        NaN        NaT
1    Batman  Batmobile 1940-04-25
2  Catwoman   Bullwhip        NaT 



In [44]:
newestdf1 = newestdf.copy()

In [45]:
newestdf.dropna(axis = 0, how = 'any', inplace = True) #removed all the missing element rows. axis = 1 for removing missing element cols
# if how was 'all' then it would remove row only if all elements of that row had missing values
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html
print(newestdf,'\n')

     name        toy       born
1  Batman  Batmobile 1940-04-25 



In [46]:
newestdf1.loc[0, 'name'] = 'Catwoman'
print(newestdf1,'\n')

       name        toy       born
0  Catwoman        NaN        NaT
1    Batman  Batmobile 1940-04-25
2  Catwoman   Bullwhip        NaT 



In [47]:
newestdf1.drop_duplicates(subset=['name'], keep = 'last', inplace = True)
# https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html
print(newestdf1,'\n')


       name        toy       born
1    Batman  Batmobile 1940-04-25
2  Catwoman   Bullwhip        NaT 



In [48]:
print(newestdf1.shape,'\n')

(2, 3) 



In [49]:
print(newestdf1.info(),'\n')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2 entries, 1 to 2
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   name    2 non-null      object        
 1   toy     2 non-null      object        
 2   born    1 non-null      datetime64[ns]
dtypes: datetime64[ns](1), object(2)
memory usage: 64.0+ bytes
None 



In [50]:
print(newestdf1['toy'].value_counts(dropna=False),'\n') #if dropna = False then it will also count the missing values and how many of them are present

Batmobile    1
Bullwhip     1
Name: toy, dtype: int64 



In [51]:
print(newestdf1['born'].value_counts(dropna=True),'\n')
print(newestdf1['born'].value_counts(dropna=False),'\n')



1940-04-25    1
Name: born, dtype: int64 

1940-04-25    1
NaT           1
Name: born, dtype: int64 



In [52]:
print(newestdf1.isnull(),'\n')
print(newestdf1.notnull(),'\n')

    name    toy   born
1  False  False  False
2  False  False   True 

   name   toy   born
1  True  True   True
2  True  True  False 



In [60]:
print(newestdf1.fillna("missing"))

       name        toy                 born
1    Batman  Batmobile  1940-04-25 00:00:00
2  Catwoman   Bullwhip              missing


In [53]:
hw = pd.DataFrame({
    'height':[170,169,162],
    'weight':[68,65,52]
})


In [54]:
print(hw,'\n')
print("mean\n",hw.mean(),'\n')
print("median\n",hw.median(),'\n')
print("mode\n",hw.mode(),'\n')
print("max\n",hw.max(),'\n')
print("min\n",hw.min(),'\n')
print("count\n",hw.count(),'\n')
print("variance\n",hw.var(),'\n')
print("standard deviation\n",hw.std(),'\n')
print("correlation\n",hw.corr(),'\n')

   height  weight
0     170      68
1     169      65
2     162      52 

mean
 height    167.000000
weight     61.666667
dtype: float64 

median
 height    169.0
weight     65.0
dtype: float64 

mode
    height  weight
0     162      52
1     169      65
2     170      68 

max
 height    170
weight     68
dtype: int64 

min
 height    162
weight     52
dtype: int64 

count
 height    3
weight    3
dtype: int64 

variance
 height    19.000000
weight    72.333333
dtype: float64 

standard deviation
 height    4.358899
weight    8.504901
dtype: float64 

correlation
           height    weight
height  1.000000  0.998058
weight  0.998058  1.000000 



In [55]:
# reading from excel file
data = pd.read_excel('data.xlsx', sheet_name='Sheet1')
print(data)

   emp id           name     hobbie
0      45         dwayne  wrestling
1      56         connor        MMA
2     109  lee chong wei  badminton


In [56]:
# writing to excel file
data.iloc[1,2] = 'trash talk'

data.to_excel('data_new.xlsx', sheet_name='Sheet1', index=True)

### Assignment:

1. Download iris dataset from kaggle. It contains 6 columns and 150 rows of 3 different sepcies of flowers.

2. Make a dataframe of the csv file. After that, make three dataframes, containing 50 rows each for all the three different species.

3. Build a new dataframe, "train_df" containing 40 rows from each of the three species dataframes. Hence containing 120 rows.

4. similarly build another dataframe, "test_df" containing the rest 10 rows from each of the three species dataframes. Hence containing 30 rows.

5. Check for any NaN or missing values in the dataframes. If present, replace them with 0 if the column type is numeric, else replace with "missing".

6. Display the column types in the dataframe

7. without using any inbuilt numpy or pandas functions, compute the mean, mode, median, variance, standard deviation, max, min of each numerical column in both train and test dataframes.

8. Without using any inbuilt numpy or pandas function, compute the correlation between each pair of columns (neglect ID and species column) in both train and test dataframes

9. In train dataset, take the two columns having the highest correlation and fit a line between them using the linear regression class of sklearn. For Example, suppose SepalLength and SepalWidth has highest correlation, then you have to predict SepalLength, given SepalWidth (this is nothing but 2D LinearRegression). After that, Plot the line alongwith the corresponding SepalLength and SepalWidth Values from the test_df dataframe.

10. Predict the value of SepalLength for every 30 values of SepalWidth in the test_df dataframe using the predicted line.

11. Compute the Mean squared error between the predictions and real sepalLengh

###### Note: it might be possible that the two highest correlating columns are not sepal length and sepal width but let's say they are "x" and "y", in which case just do the exercise with those "x" and "y" columns

12. At this point you would have predicted "x" column using "y" column and computed the error in the test_df. You have to do the same whole process, but this time you have to predict "x" column using not only "y" column but also the remaining 2 columns (whichever it be). Again compute the error. No need to plot or draw anything as it will be 4 dimensional now. 

13. Conclude which of the two methods produced less Mean square error.